### ISSUE TITLE
ansible play error AttributeError: 'NoneType' object has no attribute 'name'

### ISSUE THREAD

**Issue by raji789** [ansible/ansible#18099]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

/ansible/plugins/strategy/**init**.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

inventory = /etc/ansible/hosts

host_key_checking = False

retry_files_enabled = False 
deprecation_warnings = False
roles_path = /etc/ansible/roles
##### OS / ENVIRONMENT

Running from RHEL 6
Managing Windows 12
##### SUMMARY

We are using ansible python modules to run playbooks programmatically. We created a Django web application and on every request to server we execute playbook programmatically. This django application is running on RHEL 6  and trying to do package installations on remote RHEL6  and windows.

When running single playbook, execution is working fine without any errors.
When try to run 2 playbooks on two different remote machines, execution is failing. Sometimes both the executions sometimes either one.
##### STEPS TO REPRODUCE
1. Create a simple django based web app and run it on Rhel and make sure ansible 2.1 is installed on the same server
2.  Make sure you have 2 windows managed hosts
3. Execute a playbook to manage windows host on every request to django web application
4. Make sure you playbook is having 5 - 10 tasks 
5. Make sure your previous execution is still in progress while submitting the second request to server.
##### EXPECTED RESULTS

Expected to execute same playbook on 2 different target hosts without any issues
##### ACTUAL RESULTS

```
Traceback (most recent call last):
  File "/opt/myapp/ansible/ansibleworker.py", line 56, in submit
    results = pbex.run()
  File "/opt/myapp/venv/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
    result = self._tqm.run(play=play)
  File "/opt/myapp/venv/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
    play_return = strategy.run(iterator, play_context)
  File "/opt/myapp/venv/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 268, in run
    results += self._wait_on_pending_results(iterator)
  File "/opt/myapp/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 409, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/opt/myapp/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 271, in _process_pending_results
    self._tqm._unreachable_hosts[host.name] = True
AttributeError: 'NoneType' object has no attribute 'name'

```

**Comment by alikins** [comment]
Is there a simple playbook that can reproduce this?

Would it be possible to test this with a ansible 2.2 release candidate? There have been a lot of changes in those code paths in 2.2.

Are the playbooks being run using delegate_to?  

Does the inventory change? (are hosts added to the inventory in the playbook)

**Comment by raji789** [comment]
I am sorry i don't have a simple play book but, you can have playbook with different tasks and have includes for different playbooks.

With version 2.2 i have not tried but, i will give a try and update this ticket.

Yes, we use delegate_to

No, Inventory file created just before execution and supplied to ansible during initialization .

**Comment by jctanner** [comment]
@raji789 If you can't reproduce this with a simple playbook, you need to provide us with an example python snippet to demonstrate the issue.

needs_info

**Comment by ansibot** [comment]
@raji789 ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by raji789** [comment]
Another update is, the issue is randomly coming in different locations of ansible code. Here are another stacktrace


Traceback (most recent call last):
  File "/opt/myapp/site/mbot/ansible/ansibleworker.py", line 56, in submit
    results = pbex.run()
  File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
    result = self._tqm.run(play=play)
  File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
    play_return = strategy.run(iterator, play_context)
  File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 268, in run
    results += self._wait_on_pending_results(iterator)
  File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 409, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 279, in _process_pending_results
    self._tqm._stats.increment('ok', host.name)
AttributeError: 'NoneType' object has no attribute 'name'

One more place

Traceback (most recent call last):
  File "/home/mbot/workspace/django-mbotworker/mbot/ansible/ansibleworker.py", line 56, in submit
    results = pbex.run()
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
    result = self._tqm.run(play=play)
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
    play_return = strategy.run(iterator, play_context)
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 268, in run
    results += self._wait_on_pending_results(iterator)
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 409, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 382, in _process_pending_results
    self._variable_manager.set_host_facts(target_host, facts.copy())
  File "/home/mbot/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/vars/__init__.py", line 601, in set_host_facts
    if host.name not in self._fact_cache:
AttributeError: 'NoneType' object has no attribute 'name'

In all cases host becoming none. when 2 play books are running parallel on different hosts with 2 different inventory files

**Comment by raji789** [comment]
Another update is, the issue is randomly coming in different locations of ansible code. Here are another stacktrace

```
Traceback (most recent call last):
 File "/opt/myapp/site/app/ansible/ansibleworker.py", line 56, in submit
 results = pbex.run()
 File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
 result = self.tqm.run(play=play)
 File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
 play_return = strategy.run(iterator, play_context)
 File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 268, in run
 results += self.wait_on_pending_results(iterator)
 File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init.py", line 409, in wait_on_pending_results
 results = self.process_pending_results(iterator)
 File "/opt/myapp/venv/venv/lib/python2.7/site-packages/ansible/plugins/strategy/__init.py", line 279, in _process_pending_results
 self._tqm._stats.increment('ok', host.name)
 AttributeError: 'NoneType' object has no attribute 'name'
```

One more place

```
Traceback (most recent call last):
 File "/home/myapp/django-app/app/ansible/ansibleworker.py", line 56, in submit
 results = pbex.run()
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
 result = self.tqm.run(play=play)
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
 play_return = strategy.run(iterator, play_context)
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 268, in run
 results += self.wait_on_pending_results(iterator)
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init.py", line 409, in wait_on_pending_results
 results = self.process_pending_results(iterator)
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init.py", line 382, in process_pending_results
 self.variable_manager.set_host_facts(target_host, facts.copy())
 File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/vars/__init.py", line 601, in set_host_facts
 if host.name not in self._fact_cache:
 AttributeError: 'NoneType' object has no attribute 'name'
```

In all cases host becoming none. when 2 play books are running parallel on different hosts with 2 different inventory files

**Comment by raji789** [comment]
sample code used to run playbook attached

**Comment by raji789** [comment]
[SampleCode.txt](https://github.com/ansible/ansible/files/585403/SampleCode.txt)

**Comment by raji789** [comment]
Another update 

Updated ansible to 2.2.0.0 and issue is still happening.

```
Traceback (most recent call last):
  File "/home/app/workspace/django-appworker/app/ansible/ansibleworker.py", line 56, in submit
    results = pbex.run()
  File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 154, in run
    result = self._tqm.run(play=play)
  File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 282, in run
    play_return = strategy.run(iterator, play_context)
  File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 278, in run
    results += self._wait_on_pending_results(iterator)
  File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 534, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/home/app/.virtualenvs/worker/local/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 491, in _process_pending_results
    self._tqm._stats.increment('ok', original_host.name)
AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by raji789** [comment]
We followed python sample in the following lync to execute playbooks programatically
https://serversforhackers.com/running-ansible-programmatically

**Comment by alikins** [comment]
> We followed python sample in the following lync to execute playbooks programatically
> https://serversforhackers.com/running-ansible-programmatically

See https://serversforhackers.com/running-ansible-programmatically#comment-2554676993

That example code does not work with ansible 2.0+.
https://serversforhackers.com/running-ansible-2-programmatically is referenced as an ansible 2.0+ compatible version, thought I haven't tested it.

**Comment by raji789** [comment]
I fixed this issue my self.

We are using python threading framework and spanning multiple threads and running same playbooks on different target hosts in each thread. This approach is causing issue. 

We now used multiprocessing instead threading and we don't see the issue I stated in this ticket.

If this is something not to be fixed in ansible code go ahead and close this issue.

Thank you all for your time and support.
